File: App/go_quick_app/lib/models/thuc_pham.dart
import 'dart:convert';

import 'package:go_quick_app/models/danh_muc.dart';

List<ThucPham> listThucPhamFromJson(String str) =>
    List<ThucPham>.from(json.decode(str).map((x) => ThucPham.fromJson(x)));

String listThucPhamToJson(List<ThucPham> data) =>
    json.encode(List<dynamic>.from(data.map((x) => x.toJson())));

class ThucPham {
  ThucPham({
    this.maThucPham,
    this.ten,
    this.moTa,
    this.giaTien,
    this.urlHinhAnh,
    this.chiTiet,
    this.isDeleted,
    this.trangThai,
    this.danhMuc,
    this.createdAt,
    this.updatedAt,
  });

  int? maThucPham;
  String? ten;
  String? moTa;
  double? giaTien;
  List<String>? urlHinhAnh;
  DanhMuc? danhMuc;
  bool? isDeleted;
  String? chiTiet;
  String? trangThai;
  DateTime? createdAt;
  DateTime? updatedAt;

  factory ThucPham.fromJson(Map<String, dynamic> json) => ThucPham(
        maThucPham: json["maThucPham"],
        ten: json["ten"],
        moTa: json["moTa"],
        giaTien: json["giaTien"],
        isDeleted: json["deleted"],
        urlHinhAnh: List<String>.from(json["urlHinhAnh"].map((x) => x)),
        chiTiet: json["chiTiet"],
        danhMuc: DanhMuc.fromJson(json["danhMuc"]),
        trangThai: json["trangThai"],
        createdAt: json["createdAt"] != null
            ? DateTime.parse(json["createdAt"]).add(const Duration(hours: 7))
            : null,
        updatedAt: json["updatedAt"] != null
            ? DateTime.parse(json["updatedAt"]).add(const Duration(hours: 7))
            : null,
      );

  Map<String, dynamic> toJson() => {
        "maThucPham": maThucPham,
        "ten": ten,
        "moTa": moTa,
        "giaTien": giaTien,
        "urlHinhAnh": List<dynamic>.from(urlHinhAnh!.map((x) => x)),
        "chiTiet": chiTiet,
        "deleted": isDeleted,
        "trangThai": trangThai,
        "danhMuc": danhMuc?.toJson(),
      };
}

File: Backend/src/main/java/com/example/easynotes/model/ThucPham.java
package com.example.easynotes.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.Nationalized;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import javax.persistence.*;
import javax.validation.constraints.Min;
import javax.validation.constraints.NotBlank;
import java.io.Serializable;
import java.util.Date;
import java.util.List;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@EntityListeners(AuditingEntityListener.class)
@JsonIgnoreProperties(value = {"createdAt", "updatedAt"}, allowGetters = true)
@Entity
@Table(name = ThucPham.TABLE_NAME)
public class ThucPham implements Serializable {
    public static final String TABLE_NAME= "ThucPham";

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "maThucPham")
    private Long maThucPham;

    @Nationalized
    @NotBlank
    private String ten;

    @Nationalized
    @Lob
    @NotBlank
    @Column(name = "moTa")
    private String moTa;

    @Column(name = "daXoa")
    private boolean isDeleted;

    @Column(name = "giaTien")
    @Min(value = 0)
    private double giaTien;

    @Lob
    @ElementCollection
    private List<String> urlHinhAnh;

    @Nationalized
    @Lob
    @Column(name = "chiTiet")
    private String chiTiet;

    @Nationalized
    @NotBlank
    @Column(name = "trangThai")
    private String trangThai;

    @ManyToOne
    @JoinColumn(name = "maNhaCungCap")
    private NhaCungCap nhaCungCap;

    @ManyToOne
    @JoinColumn(name = "maDanhMuc")
    private DanhMuc danhMuc;

    @Column(nullable = false, updatable = false, name = "ngayTao")
    @Temporal(TemporalType.TIMESTAMP)
    @CreatedDate
    private Date createdAt;

    @Column(nullable = false, name = "ngayCapNhat")
    @Temporal(TemporalType.TIMESTAMP)
    @LastModifiedDate
    private Date updatedAt;







}
